src/tools/rbd/action/Ubbd.cc

```diff
@@ -0,0 +1,250 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "common/errno.h"
+#include "common/Formatter.h"
+#include "common/TextTable.h"
+#include "tools/rbd/Shell.h"
+#include "tools/rbd/Utils.h"
+
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/scope_exit.hpp>
+#include <iostream>
+#include <sys/stat.h>
+
+#if defined(WITH_RBD_UBBD)
+#include <libubbd.h>
+#endif
+
+namespace rbd {
+namespace action {
+namespace ubbd {
+
+namespace at = argument_types;
+namespace po = boost::program_options;
+
+int execute_list(const po::variables_map &vm,
+                 const std::vector<std::string> &ceph_global_init_args) {
+#if defined(WITH_RBD_UBBD)
+  ubbdd_mgmt_rsp rsp = { 0 };
+  ubbd_list_options list_opts = { .type = UBBD_DEV_TYPE_RBD };
+
+  int r = ubbd_list(&list_opts, &rsp);
+  if (r) {
```

Comment:
Hopefully `ubbd_list` returns negative errors?
```suggestion
  if (r < 0) {
```

Reply:
Actually, it returns negative error, I use "if (r)" to check it just because it can find out all non-zero ret-value, which is not expected, no matter it's positive or negative.

Reply:
This approach is potentially problematic because then one can't tell whether `r` is positive or negative (and consequently what is returned to the caller, which may be checking for `(r < 0)` or just `(r)`.  Most of the time it doesn't matter, but, in some places, errors are stashed in a data structure and/or negated so the sign matters.

Let's be consistent and use check for `(r < 0)` throughout, except where the API is expected to return a positive integer.

Reply:
ok thanx